### PR TITLE
feat(permit): refactor permit-utils package

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Refactoring permit fns into an isolated lib.

Removed all internal dependencies to have the lib isolated:
image

This lib is also to be published to npm in my personal account for now https://www.npmjs.com/package/alfetopito-permit-utils?activeTab=code

## TODO:

- [ ] Add GH action to publish libs - need devops/Anxo help with permissions, will be done in another PR

# To Test

1. Permit should work as before

### CODE REVIEW

File: libs/permit-utils/package.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "@cowprotocol/permit-utils",
+  "version": "0.0.1-RC.0",
```
Comment (alfetopito) [inline libs/permit-utils/package.json:3]: These fields define the final npm package.

File: libs/permit-utils/project.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    "build": {
+      "executor": "@nx/vite:build",
+      "outputs": ["{options.outputPath}"],
+      "options": {
+        "outputPath": "dist/libs/permit-utils"
+      }
+    },
+    "publish": {
+      "command": "node tools/scripts/publish.mjs permit-utils {args.tag} {args.otp}",
+      "dependsOn": ["build"]
```
Comment (alfetopito) [inline libs/permit-utils/project.json:16]: Command used to publish:

```bash
npx nx run permit-utils:publish [--tag <tag>] [--otp <otp>]
```

Right now published to my personal account with a different package name https://www.npmjs.com/package/alfetopito-permit-utils?activeTab=code

File: libs/permit-utils/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export { checkIsCallDataAValidPermit } from './lib/checkIsCallDataAValidPermit'
+export { generatePermitHook } from './lib/generatePermitHook'
+export { getPermitUtilsInstance } from './lib/getPermitUtilsInstance'
+export { getTokenPermitInfo } from './lib/getTokenPermitInfo'
+
+export type { PermitHookData, PermitHookParams, PermitInfo, PermitType, SupportedPermitInfo } from './types'
```
Comment (alfetopito) [inline libs/permit-utils/src/index.ts:6]: Lib interface

File: libs/permit-utils/src/lib/generatePermitHook.ts [hunk 4]
```diff
@@ -1,11 +1,9 @@
-import { GP_VAULT_RELAYER } from '@cowprotocol/common-const'
```
Comment (alfetopito) [inline libs/permit-utils/src/lib/generatePermitHook.ts:1]: Remove all internal cross dependencies

File: package.json [hunk 5]
```diff
@@ -255,6 +256,7 @@
     "ts-mockito": "^2.6.1",
     "ts-node": "^10.9.1",
     "typescript": "~5.1.3",
+    "verdaccio": "^5.0.4",
```
Comment (alfetopito) [inline package.json:259]: Dependencies added by `nx` when I ran the command to add a new lib 🤷

File: tools/scripts/publish.mjs [hunk 6]
```diff
@@ -22,16 +23,20 @@ function invariant(condition, message) {
...
+
+// Fetch the version from "package.json" before publishing
+let version
+try {
+  const json = JSON.parse(readFileSync(`package.json`).toString())
+  version = json.version
+} catch (e) {
+  console.error(chalk.bold.red(`Error reading package.json file from library build output.`))
+  process.exit(1)
+}
```
Comment (alfetopito) [inline tools/scripts/publish.mjs:38]: Changed the default publish command to use the package's version rathern than updating it based on cli param

File: tools/scripts/publish.mjs [hunk 7]
```diff
@@ -49,17 +54,10 @@ const copyReadmeCommand = `cp ${rootLib}/README.md ${outputPath}`
...
-  json.version = version
-  writeFileSync(`package.json`, JSON.stringify(json, null, 2))
-} catch (e) {
-  console.error(chalk.bold.red(`Error reading package.json file from library build output.`))
-}
+
 
 // Execute "npm publish" to publish
-const publishCommand = `npm publish --access public --tag ${tag === 'undefined' ? 'next' : tag}`
+const publishCommand = `npm publish --access public --tag ${tag === 'undefined' ? 'next' : tag} ${otp? `--otp ${otp}`:''}`
```
Comment (alfetopito) [inline tools/scripts/publish.mjs:60]: Optional `opt` param, as I needed to publish it locally under my account.
Shouldn't be needed when moving this over to CI 🤞

File: project.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  "targets": {
+    "local-registry": {
+      "executor": "@nx/js:verdaccio",
+      "options": {
+        "port": 4873,
+        "config": ".verdaccio/config.yml",
+        "storage": "tmp/local-registry/storage"
+      }
+    }
+  }
```
Comment (anxolin) [inline project.json:13]: Why did you add this file? 

It sounds good to have our own local registry, but not sure should have been part of this PR. How would like us to test it?

**Review by anxolin — APPROVED** [review]

File: package.json [hunk 9]
```diff
@@ -263,5 +265,8 @@
     "vite-plugin-svgr": "^3.2.0",
     "vite-tsconfig-paths": "~4.2.0",
     "vitest": "~0.32.0"
+  },
+  "nx": {
+    "includedScripts": []
```
Comment (shoom3301) [inline package.json:270]: Why do we need this? I remember there was a problem with it. The `nx` parameter made nx think that the root is a project

File: tools/scripts/publish.mjs [hunk 10]
```diff
@@ -7,12 +7,13 @@
  * You might need to authenticate with NPM before running this script.
  */
 
-import devkit from '@nx/devkit'
-import chalk from 'chalk'
+import devkit from "@nx/devkit";
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply tools/scripts/publish.mjs:10]: I think the default editor config for js I have is bad.
Will fix.

File: project.json (hunk 8 above)
Comment (alfetopito) [reply project.json:13]: Automatically created.
Will remove verdaccio and all of this.

File: package.json (hunk 9 above)
Comment (alfetopito) [reply package.json:270]: I think this one case when I ran `npx nx repair`.
Will remove.

File: package.json (hunk 5 above)
Comment (alfetopito) [reply package.json]: Removed

**Review by shoom3301 — APPROVED** [review]
Haven't tested, but code looks good!

File: libs/permit-utils/package.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  },
+  "module": "./index.js",
+  "main": "./index.cjs",
+  "exports": {
+    ".": {
+      "import": "./index.js",
+      "require": "./index.cjs"
+    }
+  }
+}
```
Comment (alfetopito) [inline libs/permit-utils/package.json:19]: This is so package can be consumed either the ES way (index.js) or CommonJS way (index.cjs)

File: libs/permit-utils/vite.config.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+      entry: 'src/index.ts',
+      name: 'permit-utils',
+      fileName: 'index',
+      // Change this to the formats you want to support.
+      // Don't forget to update your package.json as well.
+      formats: ['es', 'cjs'],
+    },
+    rollupOptions: {
+      // External packages that should not be bundled into your library.
+      external: [/@1inch/, /@cowprotocol/, /@ethersproject/],
```
Comment (alfetopito) [inline libs/permit-utils/vite.config.ts:49]: Explicitly removing third party dependencies to not have them included in the bundle.
They are instead added as part of package.json dependencies.
